near-sdk/src/types/public_key.rs

```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    fn from_parts(curve: CurveType, data: Vec<u8>) -> Result<Self, ParsePublicKeyError> {
+        let expected_length = match curve {
+            CurveType::ED25519 => 32,
+            CurveType::SECP256K1 => 64,
+        };
+        if data.len() != expected_length {
+            return Err(ParsePublicKeyError {
+                kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
+            });
+        }
+        let mut bytes = Vec::with_capacity(1 + expected_length);
+        bytes.push(curve as u8);
+        bytes.extend(data);
+
+        Ok(Self { data: bytes })
+    }
+
+    /// Returns a byte slice of this `PublicKey`'s contents.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+
+    /// Converts a `PublicKey` into a byte vector.
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.data
+    }
+
+    /// Get info about the CurveType for this public key
+    pub fn curve_type(&self) -> CurveType {
+        CurveType::from_u8(self.data[0]).unwrap_or_else(|_| unreachable!())
+    }
+}
+
+impl From<PublicKey> for Vec<u8> {
+    fn from(v: PublicKey) -> Vec<u8> {
+        v.data
+    }
+}
+
+impl TryFrom<Vec<u8>> for PublicKey {
+    type Error = ParsePublicKeyError;
+
+    fn try_from(mut data: Vec<u8>) -> Result<Self, Self::Error> {
+        if data.len() == 0 {
+            return Err(ParsePublicKeyError {
+                kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
+            });
+        }
+
+        let curve = CurveType::from_u8(data.remove(0))?;
+        Self::from_parts(curve, data)
+    }
+}
+
+impl serde::Serialize for PublicKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&String::from(self))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for PublicKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s: String = serde::Deserialize::deserialize(deserializer)?;
+        s.parse::<PublicKey>().map_err(serde::de::Error::custom)
+    }
+}
+
+impl From<&PublicKey> for String {
+    fn from(str_public_key: &PublicKey) -> Self {
+        match str_public_key.curve_type() {
+            CurveType::ED25519 => {
+                ["ed25519:", &bs58::encode(&str_public_key.data[1..]).into_string()].concat()
+            }
+            CurveType::SECP256K1 => {
+                ["secp256k1:", &bs58::encode(&str_public_key.data[1..]).into_string()].concat()
+            }
+        }
+    }
+}
+
+impl TryFrom<String> for PublicKey {
+    type Error = ParsePublicKeyError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_str())
+    }
+}
+
+impl TryFrom<&str> for PublicKey {
+    type Error = ParsePublicKeyError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Ok(value.parse::<Self>()?)
+    }
+}
```

Comment:
We probably don't need either of these, as I'm sure @matklad would point out, but I see some benefit in keeping so I don't have a strong opinion

Reply:
Converting from a string into a `PublicKey` is still used in `environment/env.rs` and testing, so I think we should still keep these

Reply:
These are all using `FromStr` at its base, though, so they probably should just use `FromStr` directly imo

Reply:
I see, I'll switch to that then